File: harness/blue_defend.py
"""Harness: same poisoned resume → sanitize + secure Gemini + mail interceptor."""
from __future__ import annotations

import asyncio
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "server"))

from app.agents.secure_gemini import run_secure_gemini  # noqa: E402
from app.catalog import sample_patterns  # noqa: E402
from app.config import ensure_data_dirs  # noqa: E402
from app.defense.delimiters import wrap_untrusted  # noqa: E402
from app.pdf.extract import extract_text  # noqa: E402
from app.pdf.inject import build_resume_pdf, inject_payloads  # noqa: E402
from app.pdf.sanitize import sanitize_resume_text  # noqa: E402


async def main() -> None:
    ensure_data_dirs()
    count = int(sys.argv[1]) if len(sys.argv) > 1 else 5
    base = build_resume_pdf(
        name="Blue Team Candidate",
        skills="Defensive ML, policy engines",
        experience="Hardened resume screening pipelines.",
    )
    payloads = sample_patterns(count)
    poisoned = inject_payloads(base, payloads)
    text = extract_text(poisoned["pdf_bytes"])["text"]
    sanitize = sanitize_resume_text(text)
    delimited = wrap_untrusted(sanitize["cleanedText"])
    print(f"[blue] findings={sanitize['suspectedInjectionCount']} payloads={count}")
    agent = await run_secure_gemini(delimited, payload_count=count)
    print(json.dumps({
        "mode": "blue",
        "sanitizeFindings": len(sanitize["findings"]),
        "thoughts": agent["thoughts"],
        "blockedTools": agent["blockedTools"],
        "alerts": agent["alerts"],
        "mail": agent["mailResults"],
        "modelText": agent["text"][:500],
    }, indent=2))


if __name__ == "__main__":
    asyncio.run(main())

File: harness/blue-defend.ts
/**
 * Blue harness: same poisoned PDF → sanitize + secure Gemini + interceptor alert.
 * Usage: npx tsx harness/blue-defend.ts [count=5]
 */
import dotenv from "dotenv";
import path from "node:path";
import fs from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { buildResumePdf, injectPayloads } from "../server/src/pdf/inject.js";
import { extractText } from "../server/src/pdf/extract.js";
import { sanitizeResumeText } from "../server/src/pdf/sanitize.js";
import { wrapUntrusted } from "../server/src/defense/delimiters.js";
import { samplePatterns } from "../server/src/catalog/index.js";
import { runSecureGemini } from "../server/src/agents/secureGemini.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, "../.env") });
const count = Number(process.argv[2] || 5);

process.env.MAIL_DRY_RUN ??= "true";
process.env.EMAIL_DOMAIN_WHITELIST ??= "company.com";
process.env.ADMIN_ALERT_EMAIL ??= "admin@company.com";

async function main() {
  console.log(`\n=== ResumeGuard BLUE harness (N=${count}) ===\n`);
  const base = await buildResumePdf({
    name: "Alex Candidate",
    skills: "TypeScript, Node.js",
    experience: "Built recruiting automation prototypes.",
  });
  const payloads = samplePatterns(count);
  const { pdfBytes } = await injectPayloads(base, { payloads });
  const out = path.join(__dirname, "../fixtures/resumes/poisoned-blue.pdf");
  await fs.mkdir(path.dirname(out), { recursive: true });
  await fs.writeFile(out, pdfBytes);

  const extracted = await extractText(pdfBytes);
  const sanitize = sanitizeResumeText(extracted.text);
  const delimited = wrapUntrusted(sanitize.cleanedText);
  console.log(`Sanitize findings: ${sanitize.findings.length}`);
  console.log(`Delimited preview:\n${delimited.slice(0, 300)}…\n`);

  const agent = await runSecureGemini(delimited, { payloadCount: count });

  if (agent.alerts.length === 0 && sanitize.suspectedInjectionCount > 0) {
    const { interceptSendEmail } = await import("../server/src/mail/interceptor.js");
    const probe = await interceptSendEmail(
      {
        to: "attacker@evil.example",
        subject: "Blocked probe — injection signatures in resume",
        body: `Sanitizer findings: ${sanitize.suspectedInjectionCount}`,
      },
      { payloadCount: count, toolArgs: { source: "sanitize_detector" } },
    );
    if (probe.alert) agent.alerts.push(probe.alert);
  }

  console.log("Thoughts:", agent.thoughts);
  console.log("Model:", agent.text.slice(0, 400));
  console.log("Blocked tools:", agent.blockedTools.length);
  console.log("Alerts:", JSON.stringify(agent.alerts, null, 2));
  console.log("Unauthorized mail sent:", agent.mailResults.length > 0 ? "YES (unexpected)" : "NO");

  const ok = agent.alerts.length > 0 || agent.blockedTools.length > 0 || agent.mailResults.length === 0;
  console.log(ok ? "\nBLUE SUCCESS: unauthorized send blocked / alerted" : "\nBLUE: review output");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
